recipe-server/client/control_new/components/extensions/ExtensionForm.js

```diff
@@ -0,0 +1,140 @@
+import { Button, Form, Icon, Input, Upload } from 'antd';
+import autobind from 'autobind-decorator';
+import { Map } from 'immutable';
+import React, { PropTypes as pt } from 'react';
+import { isEqual } from 'underscore';
+
+import BaseAntForm from 'control_new/components/common/BaseAntForm';
+import FormActions from 'control_new/components/common/FormActions';
+
+
+/**
+ * Form for editing XPIs uploaded to the service.
+ *
+ * The XPI field does not use normal Ant field wrappers due to how the Upload
+ * component works. Instead, the resulting file is stored in the form state and
+ * appended to the form values after validation.
+ */
+@autobind
+export class _ExtensionForm extends BaseAntForm {
+  static propTypes = {
+    extension: pt.instanceOf(Map),
+    form: pt.object,
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      xpiName: null,
+      xpiUrl: null,
+      xpiFile: null,
+    };
+  }
+
+  componentWillMount() {
+    const { extension } = this.props;
+    if (extension) {
+      this.setInitialValues(extension);
+    }
+  }
+
+  componentWillReceiveProps(newProps) {
+    // Reset form if the extension changed
+    if (!isEqual(newProps.extension, this.props.extension)) {
+      this.setInitialValues(newProps.extension);
+    }
+  }
+
+  /**
+   * Reset the form values for a new extension.
+   * @param {Object} extension
+   */
+  setInitialValues(extension) {
+    // Initial values are mostly handled via props, but if the extension
+    // changes, we need to reset the values stored in the state.
+    this.props.form.resetFields();
+    this.setState({
+      xpiName: extension.get('xpi'),
+      xpiUrl: extension.get('xpi'),
+    });
+  }
+
+  /**
+   * Handle selection of an XPI file before ant tries to upload it.
+   * @param  {File} file
+   */
+  beforeUpload(file) {
+    this.setState({
+      xpiName: file.name,
+      xpiFile: file,
+      xpiUrl: null,
+    });
+
+    // Returning false stops the Upload component from trying to send
+    // the file to the server itself.
+    return false;
+  }
+
+  getUploadFileList() {
+    const { xpiName, xpiUrl } = this.state;
+    if (xpiName) {
+      return [{
+        uid: -1,
+        name: xpiName,
+        url: xpiUrl,
+      }];
+    }
+
+    return [];
```

Comment:
To avoid having multiple `return`s, I'd suggest setting up a variable, filling if it necessary, and then returning it, e.g.:

```
let fileList = [];
if (xpiName){ fileList = [{ ... }]; }
return fileList;
```

The linter will occasionally flag `unreachable code` when there are multiple returns in a function, so I try to take this approach when possible.